### PR TITLE
Optimization over mixed spaces in `optimize_acqf_homotopy`

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

So far, `optimize_acqf_homotopy` can only handle `fixed_features` but not `fixed_features_list` which is useful for optimization over mixed spaces. In the spirit of `optimize_acqf_list`, this PR adds the option for `fixed_features_list` by using `optimize_acqf_mixed` instead of `optimize_acqf` in `optimize_acqf_homotopy`, when `fixed_features_list` is provided.

Currently, it is not working as `optimize_acqf_mixed` has no option `return_best_only=False` like `optimize_acqf`. The easiest way to solve it (maybe not the best), would be to provide an option to return the candidates from all restarts for the finally picked fixed features combination in `optimize_acqf_mixed`. For the start, it would be sufficient to only implement it for `q=1`, and error out for the case of `q>1`.

What do you think?

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

### CODE REVIEW

File: botorch/optim/optimize_homotopy.py
Comment (Balandat) [inline botorch/optim/optimize_homotopy.py:166]: `fixed_features` as an arg seems redundant if we have `fixed_features_list`. We should probably deprecate the  `fixed_features` arg and raise a warning when passed. And if `fixed_features_list` contains only one element we can just use that with `optimize_acqf`

File: test/optim/test_homotopy.py
Comment (Balandat) [inline test/optim/test_homotopy.py:155]:
```suggestion
            "Either `fixed_feature` or `fixed_features_list` can be provided, not both.",
```

**Comment by Balandat** [comment]
> Currently, it is not working as optimize_acqf_mixed has no option return_best_only=False like optimize_acqf. The easiest way to solve it (maybe not the best), would be to provide an option to return the candidates from all restarts for the finally picked fixed features combination in optimize_acqf_mixed. For the start, it would be sufficient to only implement it for q=1, and error out for the case of q>1. What do you think?

I think that makes sense. Starting with only supporting the `q=1` case seems very reasonable (let's add the idea above in the comments though where we raise that `UnsupportedError`).

**Comment by jduerholt** [comment]
I added the possibility `return_best_only=False` for `q=1` in `optimize_acqf_mixed` and adapted the tests. Also `optimize_acqf_homotopy` works now with a fixed_features_list. For this, I had to remove the following options from `optimize_acqf_homotopy` as they are not available for `optimize_acqf_mixed`:
- `gen_candidates`: would it make sense for you to add this option to `optimize_acqf_mixed`?
- `sequential`: this keyword make for me no sense in the kontext of `optimize_homotopy`, as we are always using `q=1` for the calls to `optimize_acqf`
 - `return_full_tree`: I am not sure about this as I do not exactly know what it does. Does it make sense in kontext of `optimize_homotopy`?

What do you think?

File: botorch/optim/optimize_homotopy.py
Comment (Balandat) [inline botorch/optim/optimize_homotopy.py:176]: yup, makes sense

File: botorch/optim/optimize_homotopy.py
Comment (Balandat) [reply botorch/optim/optimize_homotopy.py]: yes, that would be great

File: botorch/optim/optimize_homotopy.py
Comment (Balandat) [inline botorch/optim/optimize_homotopy.py:179]: No we can probably just remove this from `optimize_acqf_homotopy`

File: botorch/optim/optimize_homotopy.py
Comment (Balandat) [inline botorch/optim/optimize_homotopy.py:242]: You can make this more concise  - you should define `optimization_fn` and `fixed_feature_kwargs` above before the `for _ in range(q):`  block so you can reuse this below as well (but GH doesn't let me make that code suggestion). 
```suggestion
        if fixed_features_list:
            optimization_fn = optimize_acqf_mixed
            fixed_feature_kwargs = {"fixed_features_list": fixed_features_list}
        else:
            optimization_fn = optimize_acqf
            fixed_feature_kwargs = {"fixed_features": fixed_features}
        
        candidates, acq_values = optimization_fn(
                acq_function=acq_function,
                bounds=bounds,
                q=1,
                options=final_options,
                batch_initial_conditions=candidates,
                **fixed_feature_kwargs,
                **shared_optimize_acqf_kwargs,
            )
```

**Review by Balandat — COMMENTED** [review]
> gen_candidates: would it make sense for you to add this option to optimize_acqf_mixed?

Yeah I think that would make sense. Inside of `optimize_acqf_mixed` we'd just pass this to `optimize_acqf`.

> sequential: this keyword make for me no sense in the kontext of optimize_homotopy, as we are always using q=1 for the calls to optimize_acqf

True - I think we can remove this from `optimize_acqf_homotopy`.

> return_full_tree: I am not sure about this as I do not exactly know what it does. Does it make sense in kontext of optimize_homotopy?

This is used by lookahead acquisition functions. I don't think we need to support this for `optimize_acqf_homotopy`.

File: botorch/optim/optimize.py
Comment (Balandat) [reply botorch/optim/optimize.py:975]: Did you? I don't see this change.

Comment (jduerholt) [reply botorch/optim/optimize.py:975]: Yeah, I only did it locally. It is kind of a todolist for me here, with this comment it was for me "abgehakt" :D Will push all at once later ;)

File: botorch/optim/optimize.py
Comment (Balandat) [inline botorch/optim/optimize.py:985]: Can you move this before the `for` loop? Don't need to redo this each iteration and it will also make things more readable.

File: botorch/optim/optimize_homotopy.py
Comment (Balandat) [inline botorch/optim/optimize_homotopy.py:212]: I don't think you want to make things conditional on this - we did this even before supporting `fixed_features_list`

File: test/optim/test_homotopy.py
Comment (Balandat) [inline test/optim/test_homotopy.py:170]:
```suggestion
            bounds=torch.tensor([[-10, -10], [5, 5]], **tkwargs),
```

File: test/optim/test_homotopy.py
Comment (Balandat) [inline test/optim/test_homotopy.py:160]:
```suggestion
                bounds=torch.tensor([[-10, -10], [5, 5]], **tkwargs),
```

File: test/optim/test_optimize.py
Comment (Balandat) [inline test/optim/test_optimize.py:1479]:
```suggestion
                self.assertEqual(expected_candidates.dim(), 2)
```

File: test/optim/test_optimize.py
Comment (Balandat) [inline test/optim/test_optimize.py:1484]:
```suggestion
                self.assertEqual(expected_candidates.dim(), 3)
                self.assertEqual(expected_acq_value.dim(), 1)
```

File: test/optim/test_optimize.py
Comment (Balandat) [inline test/optim/test_optimize.py:1605]: nit: isolate the expected Exception in the context manager
```suggestion
        mock_acq_function = MockAcquisitionFunction()
        with self.assertRaises(
            NotImplementedError,
            msg="`return_best_only=False` is only supported for q=1.",
        ):
```

File: test/optim/test_optimize.py
Comment (Balandat) [inline test/optim/test_optimize.py:1551]: Don't think this `to` is needed here
```suggestion
                )
```

File: test/optim/test_optimize.py
Comment (Balandat) [inline test/optim/test_optimize.py:1470]: This shouldn't be needed here?
```suggestion
            )
```

**Review by Balandat — COMMENTED** [review]
Thanks - largely nits except for the conditional when "Optimize one more time with the final options".

Also looks like you have a `KeyError` test failure as well as line length lint issue

File: botorch/optim/optimize_homotopy.py
Comment (jduerholt) [reply botorch/optim/optimize_homotopy.py:229]: Good catch. Thanks! I overlooked the if when I was changing it to `optimization_fn`.

File: botorch/optim/optimize_homotopy.py
Comment (jduerholt) [reply botorch/optim/optimize_homotopy.py]: I implemeted it now for `optimize_acqf_homotopy`, if it is fine for you how it looks there, I will port it to `optimize_acqf_list`.

**Comment by jduerholt** [comment]
Hi @Balandat,

thanks for your review(s). I think, I have now implemented all the suggestions.

Could you have a look on the issue that I always have too wait for approval before the pipelines start? In the past, they were starting always automatically which helped me to faster finish the PRs ;) And maybe it is also less annoying for you ...

Best,

Johannes

**Comment by Balandat** [comment]
> Could you have a look on the issue that I always have to wait for approval before the pipelines start? In the past, they were starting always automatically which helped me to faster finish the PRs ;) And maybe it is also less annoying for you ...

Yeah I pinged some folks to understand if we can avoid this, let's see what they say.

**Comment by Balandat** [comment]
Looks like you're missing a unit test for more than 1 elements in the `fixed_features_list` which seems important :)

**Comment by jduerholt** [comment]
> Looks like you're missing a unit test for more than 1 elements in the `fixed_features_list` which seems important :)

Ah, this is now a consequence of the new structure where we call `optimize_acqf` when `fixed_features_list` has only one element. Before it was also in this case going to `òptimize_acqf_mixed`. I will add a new test case.

**Comment by jduerholt** [comment]
Now, the test is there ;)

**Review by Balandat — APPROVED** [review]
Tutorial failure is unrelated (and will be fixed in #2641). Thanks!

File: botorch/optim/optimize.py
Comment (esantorella) [inline botorch/optim/optimize.py:982]: Why not `return_best_only=return_best_only`?

File: botorch/optim/optimize.py
Comment (esantorella) [inline botorch/optim/optimize.py:1046]: Why True here and False earlier?

File: test/optim/test_homotopy.py
Comment (esantorella) [inline test/optim/test_homotopy.py:140]: Thanks for the tests!

File: test/optim/test_optimize.py
Comment (esantorella) [inline test/optim/test_optimize.py:1589]:
```suggestion
        with self.assertRaisesRegex(
```
`assertRaises` with `msg` doesn't actually check message equality -- the msg keyword is for the error message for a failure. See here: https://docs.python.org/3/library/unittest.html#unittest.TestCase.debug

File: test/optim/test_optimize.py
Comment (esantorella) [inline test/optim/test_optimize.py:1590]:
```suggestion
            ValueError, "fixed_features_list must be non-empty."
```

File: test/optim/test_optimize.py
Comment (esantorella) [inline test/optim/test_optimize.py:1606]:
```suggestion
        with self.assertRaisesRegex(
            NotImplementedError,
            regex="`return_best_only=False` is only supported for q=1.",
```

File: botorch/optim/optimize.py
Comment (jduerholt) [reply botorch/optim/optimize.py:982]: We get here all candidates and then filter later based on the `return_best_only` argument. I will add a comment.

File: botorch/optim/optimize.py
Comment (jduerholt) [reply botorch/optim/optimize.py:1046]: Because, currently `return_best_only` is only implemented for `q=1` and here we are in the loop for `q>1`.

**Comment by jduerholt** [comment]
Thanks for the review @esantorella. I adressed your comments and resolved the merge conflicts with main. Best, Johannes

**Pull request closed without merging**
